File: src/treemapper/diffctx/edges/semantic/protobuf.py
from __future__ import annotations

import re
from collections import defaultdict
from pathlib import Path

from ...config.weights import EDGE_WEIGHTS
from ...types import Fragment, FragmentId
from ..base import EdgeBuilder, EdgeDict, FragmentIndex, discover_files_by_refs

_PROTO_EXTS = {".proto"}

_PROTO_IMPORT_RE = re.compile(r'^\s{0,20}import\s{1,10}(?:public\s{1,10})?"([^"]{1,300})"\s*;', re.MULTILINE)
_PROTO_PACKAGE_RE = re.compile(r"^\s{0,20}package\s{1,10}([a-zA-Z_][\w.]{0,200})\s*;", re.MULTILINE)
_PROTO_MESSAGE_RE = re.compile(r"^\s{0,20}message\s{1,10}([A-Z]\w{0,100})\s*\{", re.MULTILINE)
_PROTO_ENUM_RE = re.compile(r"^\s{0,20}enum\s{1,10}([A-Z]\w{0,100})\s*\{", re.MULTILINE)
_PROTO_SERVICE_RE = re.compile(r"^\s{0,20}service\s{1,10}([A-Z]\w{0,100})\s*\{", re.MULTILINE)
_PROTO_RPC_RE = re.compile(
    r"^\s{0,20}rpc\s{1,10}(\w{1,100})\s*\(\s*(?:stream\s+)?([A-Z][\w.]{0,200})\s*\)"
    r"\s*returns\s*\(\s*(?:stream\s+)?([A-Z][\w.]{0,200})\s*\)",
    re.MULTILINE,
)
_PROTO_FIELD_TYPE_RE = re.compile(
    r"^\s{0,20}(?:repeated\s{1,10}|optional\s{1,10}|required\s{1,10})?"
    r"(?:map\s*<\s*\w+\s*,\s*)?([A-Z][\w.]{0,200})\s*>?\s+\w{1,100}\s*=",
    re.MULTILINE,
)
_PROTO_GO_PACKAGE_RE = re.compile(r'^\s{0,20}option\s{1,10}go_package\s*=\s*"([^"]{1,300})"\s*;', re.MULTILINE)
_PROTO_JAVA_PACKAGE_RE = re.compile(r'^\s{0,20}option\s{1,10}java_package\s*=\s*"([^"]{1,300})"\s*;', re.MULTILINE)


def _is_proto_file(path: Path) -> bool:
    return path.suffix.lower() in _PROTO_EXTS


def _extract_imports(content: str) -> set[str]:
    return {m.group(1) for m in _PROTO_IMPORT_RE.finditer(content)}


def _extract_package(content: str) -> str | None:
    m = _PROTO_PACKAGE_RE.search(content)
    return m.group(1) if m else None


def _extract_message_names(content: str) -> set[str]:
    names: set[str] = set()
    names.update(m.group(1) for m in _PROTO_MESSAGE_RE.finditer(content))
    names.update(m.group(1) for m in _PROTO_ENUM_RE.finditer(content))
    return names


def _extract_service_names(content: str) -> set[str]:
    return {m.group(1) for m in _PROTO_SERVICE_RE.finditer(content)}


def _extract_rpc_types(content: str) -> set[str]:
    types: set[str] = set()
    for m in _PROTO_RPC_RE.finditer(content):
        types.add(m.group(2).split(".")[-1])
        types.add(m.group(3).split(".")[-1])
    return types


def _extract_field_types(content: str) -> set[str]:
    types: set[str] = set()
    for m in _PROTO_FIELD_TYPE_RE.finditer(content):
        raw = m.group(1)
        types.add(raw.split(".")[-1])
    return types


_PROTO_BUILTIN_TYPES = frozenset(
    {
        "Any",
        "Timestamp",
        "Duration",
        "Empty",
        "FieldMask",
        "Struct",
        "Value",
        "ListValue",
        "BoolValue",
        "BytesValue",
        "DoubleValue",
        "FloatValue",
        "Int32Value",
        "Int64Value",
        "StringValue",
        "UInt32Value",
        "UInt64Value",
    }
)


def _extract_option_paths(content: str) -> set[str]:
    refs: set[str] = set()
    for m in _PROTO_GO_PACKAGE_RE.finditer(content):
        pkg = m.group(1)
        last_segment = pkg.rstrip("/").split("/")[-1]
        if ";" in last_segment:
            last_segment = last_segment.split(";")[0].split("/")[-1]
        refs.add(last_segment)
    for m in _PROTO_JAVA_PACKAGE_RE.finditer(content):
        pkg = m.group(1)
        last_segment = pkg.split(".")[-1]
        refs.add(last_segment)
    return refs


def _collect_proto_refs(proto_files: list[Path]) -> set[str]:
    refs: set[str] = set()
    for pf in proto_files:
        try:
            content = pf.read_text(encoding="utf-8")
        except (OSError, UnicodeDecodeError):
            continue
        refs.update(_extract_imports(content))
        refs.update(_extract_option_paths(content))
    return refs


class ProtobufEdgeBuilder(EdgeBuilder):
    weight = 0.65
    import_weight = EDGE_WEIGHTS["proto_import"].forward
    message_ref_weight = EDGE_WEIGHTS["proto_message_ref"].forward
    service_rpc_weight = EDGE_WEIGHTS["proto_service_rpc"].forward
    reverse_weight_factor = EDGE_WEIGHTS["proto_import"].reverse_factor

    def discover_related_files(
        self,
        changed_files: list[Path],
        all_candidate_files: list[Path],
        repo_root: Path | None = None,
    ) -> list[Path]:
        proto_changed = [f for f in changed_files if _is_proto_file(f)]
        if not proto_changed:
            return []

        refs = _collect_proto_refs(proto_changed)
        return discover_files_by_refs(refs, changed_files, all_candidate_files, repo_root)

    def build(self, fragments: list[Fragment], repo_root: Path | None = None) -> EdgeDict:
        proto_frags = [f for f in fragments if _is_proto_file(f.path)]
        if not proto_frags:
            return {}

        edges: EdgeDict = {}
        idx = FragmentIndex(fragments, repo_root)
        message_defs = self._build_message_index(proto_frags)

        for pf in proto_frags:
            self._add_import_edges(pf, idx, edges)
            self._add_type_ref_edges(pf, message_defs, edges)
            self._add_option_path_edges(pf, idx, edges)

        return edges

    def _build_message_index(self, proto_frags: list[Fragment]) -> dict[str, list[FragmentId]]:
        defs: dict[str, list[FragmentId]] = defaultdict(list)
        for f in proto_frags:
            pkg = _extract_package(f.content)
            for name in _extract_message_names(f.content):
                defs[name.lower()].append(f.id)
                if pkg:
                    defs[f"{pkg}.{name}".lower()].append(f.id)
            for name in _extract_service_names(f.content):
                defs[name.lower()].append(f.id)
                if pkg:
                    defs[f"{pkg}.{name}".lower()].append(f.id)
        return defs

    def _add_import_edges(self, pf: Fragment, idx: FragmentIndex, edges: EdgeDict) -> None:
        for imp in _extract_imports(pf.content):
            self.link_by_path_match(pf.id, imp, idx, edges, self.import_weight)

    def _add_type_ref_edges(self, pf: Fragment, message_defs: dict[str, list[FragmentId]], edges: EdgeDict) -> None:
        own_defs = _extract_message_names(pf.content)
        own_defs_lower = {n.lower() for n in own_defs}

        field_types = _extract_field_types(pf.content)
        rpc_types = _extract_rpc_types(pf.content)

        for type_name in field_types:
            if type_name in _PROTO_BUILTIN_TYPES:
                continue
            type_lower = type_name.lower()
            if type_lower in own_defs_lower:
                continue
            for fid in message_defs.get(type_lower, []):
                if fid != pf.id:
                    self.add_edge(edges, pf.id, fid, self.message_ref_weight)

        for type_name in rpc_types:
            if type_name in _PROTO_BUILTIN_TYPES:
                continue
            type_lower = type_name.lower()
            if type_lower in own_defs_lower:
                continue
            for fid in message_defs.get(type_lower, []):
                if fid != pf.id:
                    self.add_edge(edges, pf.id, fid, self.service_rpc_weight)

    def _add_option_path_edges(self, pf: Fragment, idx: FragmentIndex, edges: EdgeDict) -> None:
        for ref in _extract_option_paths(pf.content):
            self.link_by_path_match(pf.id, ref, idx, edges, self.weight * 0.6)

File: diffctx/src/edges/semantic/protobuf.rs
use std::path::{Path, PathBuf};

use once_cell::sync::Lazy;
use regex::Regex;
use rustc_hash::{FxHashMap, FxHashSet};

use crate::config::weights::EDGE_WEIGHTS;
use crate::types::Fragment;

use super::super::EdgeDict;
use super::super::base::{self, EdgeBuilder, add_edge, discover_files_by_refs};

fn is_proto_file(path: &Path) -> bool {
    base::file_ext(path) == ".proto"
}

static IMPORT_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r#"(?m)^\s*import\s+(?:public\s+)?["']([^"']+)["']"#).unwrap());
static MSG_RE: Lazy<Regex> =
    Lazy::new(|| Regex::new(r"(?m)^\s*(?:message|enum|service)\s+(\w+)").unwrap());
static FIELD_TYPE_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?m)^\s*(?:repeated\s+|optional\s+|required\s+|map<\w+,\s*)?([A-Z]\w+)").unwrap()
});
static RPC_TYPE_RE: Lazy<Regex> = Lazy::new(|| {
    Regex::new(r"(?:returns\s*\(|rpc\s+\w+\s*\()\s*(?:stream\s+)?([A-Z]\w+)").unwrap()
});

fn extract_imports(content: &str) -> FxHashSet<String> {
    IMPORT_RE
        .captures_iter(content)
        .map(|c| c[1].to_string())
        .collect()
}

fn extract_defs(content: &str) -> FxHashSet<String> {
    MSG_RE
        .captures_iter(content)
        .map(|c| c[1].to_string())
        .collect()
}

fn extract_type_refs(content: &str) -> FxHashSet<String> {
    let mut refs: FxHashSet<String> = FIELD_TYPE_RE
        .captures_iter(content)
        .map(|c| c[1].to_string())
        .collect();
    refs.extend(RPC_TYPE_RE.captures_iter(content).map(|c| c[1].to_string()));
    refs
}

pub struct ProtobufEdgeBuilder;

impl EdgeBuilder for ProtobufEdgeBuilder {
    fn build(&self, fragments: &[Fragment], repo_root: Option<&Path>) -> EdgeDict {
        let frags: Vec<&Fragment> = fragments
            .iter()
            .filter(|f| is_proto_file(Path::new(f.path())))
            .collect();
        if frags.is_empty() {
            return FxHashMap::default();
        }

        let import_w = EDGE_WEIGHTS["proto_import"].forward;
        let msg_w = EDGE_WEIGHTS["proto_message_ref"].forward;
        let rpc_w = EDGE_WEIGHTS["proto_service_rpc"].forward;
        let reverse_factor = EDGE_WEIGHTS["proto_import"].reverse_factor;

        let idx = base::FragmentIndex::new(fragments, repo_root);
        let mut name_to_defs: FxHashMap<String, Vec<_>> = FxHashMap::default();
        for f in &frags {
            for name in extract_defs(&f.content) {
                name_to_defs
                    .entry(name.to_lowercase())
                    .or_default()
                    .push(f.id.clone());
            }
        }

        let mut edges: EdgeDict = FxHashMap::default();

        for f in &frags {
            let self_defs = extract_defs(&f.content);
            for imp in extract_imports(&f.content) {
                base::link_by_name(&f.id, &imp, &idx, &mut edges, import_w, reverse_factor);
            }
            for tref in extract_type_refs(&f.content) {
                if self_defs.contains(&tref) {
                    continue;
                }
                let w = if RPC_TYPE_RE.captures_iter(&f.content).any(|c| c[1] == tref) {
                    rpc_w
                } else {
                    msg_w
                };
                if let Some(targets) = name_to_defs.get(&tref.to_lowercase()) {
                    for t in targets {
                        if t != &f.id {
                            add_edge(&mut edges, &f.id, t, w, reverse_factor);
                        }
                    }
                }
            }
        }
        edges
    }

    fn discover_related_files(
        &self,
        changed: &[PathBuf],
        candidates: &[PathBuf],
        repo_root: Option<&Path>,
        file_cache: Option<&FxHashMap<PathBuf, String>>,
    ) -> Vec<PathBuf> {
        let proto_changed: Vec<&PathBuf> = changed.iter().filter(|f| is_proto_file(f)).collect();
        if proto_changed.is_empty() {
            return vec![];
        }
        let mut refs = FxHashSet::default();
        for f in &proto_changed {
            if let Some(content) = base::read_file_cached(f, file_cache) {
                refs.extend(extract_imports(&content));
            }
        }
        discover_files_by_refs(&refs, changed, candidates, repo_root)
    }
}
